pkg/neonvm/controllers/vm_cert_controller.go

```diff
@@ -0,0 +1,341 @@
+	log := log.FromContext(ctx)
+
+	certSecret, err := r.certSecretForVirtualMachine(vm, key, certificateReq.Status.Certificate)
+	if err != nil {
+		log.Error(err, "Failed to define new TLS secret resource for VirtualMachine")
+		return nil, err
+	}
+
+	if err = r.Create(ctx, certSecret); err != nil {
+		log.Error(err, "Failed to create new TLS secret", "Secret.Namespace", certSecret.Namespace, "Secret.Name", certSecret.Name)
+		return nil, err
+	}
+	log.Info("Virtual Machine TLS secret was created", "Secret.Namespace", certSecret.Namespace, "Secret.Name", certSecret.Name)
+
+	msg := fmt.Sprintf("VirtualMachine %s created TLS secret %s", vm.Name, certSecret.Name)
+	r.Recorder.Event(vm, "Normal", "Created", msg)
+
+	return certSecret, nil
+}
+
+func (r *VMReconciler) updateTlsSecret(ctx context.Context, vm *vmv1.VirtualMachine, key crypto.Signer, certificateReq *certv1.CertificateRequest, certSecret *corev1.Secret) error {
+	log := log.FromContext(ctx)
+
+	encodedKey, err := pki.EncodePrivateKey(key, certv1.PKCS1)
+	if err != nil {
+		return err
+	}
+	certSecret.Data[corev1.TLSPrivateKeyKey] = encodedKey
+	certSecret.Data[corev1.TLSCertKey] = certificateReq.Status.Certificate
+
+	if err = r.Update(ctx, certSecret); err != nil {
+		log.Error(err, "Failed to update new TLS secret", "Secret.Namespace", certSecret.Namespace, "Secret.Name", certSecret.Name)
+		return err
+	}
+	log.Info("Virtual Machine TLS secret was updated", "Secret.Namespace", certSecret.Namespace, "Secret.Name", certSecret.Name)
+
+	msg := fmt.Sprintf("VirtualMachine %s updated TLS secret %s", vm.Name, certSecret.Name)
+	r.Recorder.Event(vm, "Normal", "Updated", msg)
+
+	return nil
+}
+
+func (r *VMReconciler) deleteTmpSecret(ctx context.Context, vm *vmv1.VirtualMachine, tmpKeySecret *corev1.Secret, certificateReq *certv1.CertificateRequest) {
+	log := log.FromContext(ctx)
+
+	err := r.Delete(ctx, tmpKeySecret)
+	if err != nil {
+		log.Info("Virtual Machine temporary TLS private key secret could not be deleted", "Secret.Namespace", tmpKeySecret.Namespace, "Secret.Name", tmpKeySecret.Name)
+	}
+	msg := fmt.Sprintf("VirtualMachine %s temporary TLS private key secret %s was deleted", vm.Name, tmpKeySecret.Name)
+	r.Recorder.Event(vm, "Normal", "Deleted", msg)
+
+	err = r.Delete(ctx, certificateReq)
+	if err != nil {
+		log.Info("Virtual Machine CertificateRequest could not be deleted", "CertificateRequest.Namespace", certificateReq.Namespace, "CertificateRequest.Name", certificateReq.Name)
+	}
+	msg = fmt.Sprintf("VirtualMachine %s CertificateRequest %s was deleted", vm.Name, certificateReq.Name)
+	r.Recorder.Event(vm, "Normal", "Deleted", msg)
+}
+
+func certSpecCSR(vm *vmv1.VirtualMachine) (*x509.CertificateRequest, error) {
+	certSpec := certv1.CertificateSpec{
+		CommonName: vm.Spec.TLS.ServerName,
+		DNSNames:   []string{vm.Spec.TLS.ServerName},
+		PrivateKey: &certv1.CertificatePrivateKey{
+			// TODO: can we support Ed25519?
```

Comment:
question: What's the context on this TODO? will it be logged as an issue somewhere? if no, let's remove?

Reply:
Removed